FirebaseAppCheck/Sources/AppAttestProvider/API/FIRAppAttestAPIService.m

```diff
@@ -56,8 +59,83 @@ - (instancetype)initWithAPIService:(id<FIRAppCheckAPIServiceProtocol>)APIService
+  NSURL *URL = [NSURL URLWithString:URLString];
+
+  return [FBLPromise onQueue:[self defaultQueue]
+                          do:^id _Nullable {
+                            return [self.APIService sendRequestWithURL:URL
+                                                            HTTPMethod:@"POST"
+                                                                  body:nil
+                                                     additionalHeaders:nil];
+                          }]
+      .then(^id _Nullable(GULURLSessionDataResponse *_Nullable response) {
+        return [self randomChallengeWithAPIResponse:response];
+      });
+}
+
+- (FBLPromise<NSData *> *)randomChallengeWithAPIResponse:(GULURLSessionDataResponse *)response {
+  return [FBLPromise onQueue:[self defaultQueue] do:^id _Nullable{
+    NSError *error;
+
+    NSData *randomChallenge = [self randomChallengeFromResponseBody:response.HTTPBody error:&error];
+
+    return randomChallenge ?: error;
+  }];
+}
+
+- (NSData *)randomChallengeFromResponseBody:(NSData *)response
+                                      error:(NSError **)outError {
+  if (response.length <= 0) {
+    FIRAppCheckSetErrorToPointer(
+        [FIRAppCheckErrorUtil errorWithFailureReason:@"Empty server response body."], outError);
+    return nil;
+  }
+
+  NSError *JSONError;
+  NSDictionary *responseDict = [NSJSONSerialization JSONObjectWithData:response
+                                                               options:0
+                                                                 error:&JSONError];
+
+  if (![responseDict isKindOfClass:[NSDictionary class]]) {
+    FIRAppCheckSetErrorToPointer([FIRAppCheckErrorUtil JSONSerializationError:JSONError], outError);
+    return nil;
+  }
+
+  NSString *challenge = responseDict[@"challenge"];
+  if (![challenge isKindOfClass:[NSString class]]) {
+    FIRAppCheckSetErrorToPointer(
+        [FIRAppCheckErrorUtil appCheckTokenResponseErrorWithMissingField:@"challenge"],
+        outError);
+    return nil;
+  }
+
+  NSString *timeToLiveString = responseDict[@"ttl"];
+  if (![challenge isKindOfClass:[NSString class]] || challenge.length <= 0) {
+    FIRAppCheckSetErrorToPointer(
+        [FIRAppCheckErrorUtil appCheckTokenResponseErrorWithMissingField:@"ttl"], outError);
+    return nil;
+  }
+
+  // Expect a string like "3600s" representing a time interval in seconds.
+  NSString *timeToLiveValueString = [timeToLiveString stringByReplacingOccurrencesOfString:@"s"
+                                                                                withString:@""];
+  NSTimeInterval secondsToLive = timeToLiveValueString.doubleValue;
+
+  if (secondsToLive == 0) {
+    FIRAppCheckSetErrorToPointer(
+        [FIRAppCheckErrorUtil appCheckTokenResponseErrorWithMissingField:@"ttl"], outError);
+    return nil;
+  }
+
+  NSData *randomChallenge = [challenge dataUsingEncoding:NSUTF8StringEncoding];
```

Comment:
I think it's base64URL (don't be confused with just base64). Could you please check and modify if needed.

Reply:
I checked and my understanding is that `getRandomChallenge` returns `NSUTF8StringEncoding` encoded data. See it mocked as such in tests [here](https://github.com/firebase/firebase-ios-sdk/blob/f2b77f0dce2def4037806b705c7608cfb4e24298/FirebaseAppCheck/Tests/Unit/AppAttestProvider/FIRAppAttestProviderTests.m#L140)